crates/uplc/src/ast.rs

```diff
@@ -14,6 +14,32 @@ pub struct Program<T> {
     pub term: Term<T>,
 }
 
+impl<T> Program<T>
+where
+    T: Clone,
+{
+    /// We use this to apply the validator to Datum,
+    /// then redeemer, then ScriptContext. If datum is
+    /// even necessary (i.e. minting policy).
+    ///
+    /// ```rust
+    /// program.apply(&datum)
+    ///    .apply(&redeemer)
+    ///    .apply(&script_context);
+    /// ```
+    pub fn apply(&self, program: &Self) -> Self {
```

Comment:
It's weird to me that apply takes a program instead of a term, but that is indeed how it's written in the `Plutus` repo.